server/driver/wivrn_fb_face2_tracker.cpp

```diff
@@ -0,0 +1,150 @@
+
+#include "os/os_time.h"
+#include "wivrn_packets.h"
+#include "xrt/xrt_defines.h"
+#include "xrt/xrt_device.h"
+
+#include "util/u_logging.h"
+
+#include "xrt/xrt_results.h"
+#include <cmath>
+#include <cstring>
+#include <openxr/openxr.h>
+
+static void wivrn_fb_face2_tracker_destroy(xrt_device * xdev);
+
+static void wivrn_fb_face2_tracker_update_inputs(xrt_device * xdev);
+
+static xrt_result_t wivrn_fb_face2_tracker_get_face_tracking(struct xrt_device * xdev,
+                                                             enum xrt_input_name facial_expression_type,
+                                                             struct xrt_facial_expression_set * out_value);
+
+wivrn_fb_face2_tracker::wivrn_fb_face2_tracker(xrt_device * hmd,
+                                               std::shared_ptr<xrt::drivers::wivrn::wivrn_session> cnx) :
+        xrt_device{}, cnx(cnx)
+{
+	xrt_device * base = this;
+	base->tracking_origin = hmd->tracking_origin;
+	base->get_face_tracking = wivrn_fb_face2_tracker_get_face_tracking;
+	base->update_inputs = wivrn_fb_face2_tracker_update_inputs;
+	base->destroy = wivrn_fb_face2_tracker_destroy;
+	name = XRT_DEVICE_FB_FACE_TRACKING2;
+	device_type = XRT_DEVICE_TYPE_FACE_TRACKER;
+	face_tracking_supported = true;
+
+	// Print name.
+	strcpy(str, "WiVRn FB v2 Face Tracker");
+	strcpy(serial, "WiVRn FB v2 Face Tracker");
+
+	// Setup input.
+	face_input.name = XRT_INPUT_FB_FACE_TRACKING2_VISUAL;
+	face_input.active = true;
+	inputs = &face_input;
+	input_count = 1;
+}
+
+void wivrn_fb_face2_tracker::update_inputs()
+{
+	// Empty
+}
+
+void wivrn_fb_face2_tracker::update_tracking(const from_headset::fb_face2 & face, const clock_offset & offset)
+{
+	is_valid = face.is_valid;
+	is_eye_following_blendshapes_valid = face.is_eye_following_blendshapes_valid;
+	if (!face.is_valid)
+	{
+		return;
+	}
+
+	if (face.production_timestamp != face.timestamp)
+	{
+		last_prediction_offset = face.timestamp - face.production_timestamp;
+	}
+
+	weights.update_tracking(face.production_timestamp, face.timestamp, face.weights, offset);
+	confidences.update_tracking(face.production_timestamp, face.timestamp, face.confidences, offset);
+	last_received_ns = os_monotonic_get_ns();
+}
+
+xrt_result_t wivrn_fb_face2_tracker::get_face_tracking(enum xrt_input_name facial_expression_type, struct xrt_facial_expression_set * out_value)
+{
+	std::lock_guard lock(mutex);
+	if (facial_expression_type == XRT_INPUT_FB_FACE_TRACKING2_VISUAL)
+	{
+		auto now = os_monotonic_get_ns();
+
+		if ((now - last_received_ns) > 250'000'000)
+		{ // data is stale
+			is_valid = false;
+		}
+
+		out_value->face_expression_set2_fb.is_valid = is_valid;
+
+		if (!is_valid)
+		{
+			return XRT_SUCCESS;
+		}
+
+		out_value->face_expression_set2_fb.is_eye_following_blendshapes_valid = is_eye_following_blendshapes_valid;
+
+		XrTime predicted_time = ((XrTime)now) + last_prediction_offset;
+
+		out_value->face_expression_set2_fb.sample_time_ns = predicted_time;
```

Comment:
The sample_time could be stored as data with the weights list, where interpolate/extrapolate would likely take the max value. No need to store it out of the history object.

I don't if there is a reason to split weights and confidences

Reply:
the idea was to keep weights_list generic enough so that it could be used for a different face tracking extension down the line.

Reply:
If you think it makes sense to have 2 objects, it's not a big problem, but the sample time should be part of the data